transports/websocket-websys/src/web_context.rs

```diff
@@ -1,57 +1,185 @@
-                    fn worker(this: &Global) -> JsValue;
-                }
-                let global: Global = js_sys::global().unchecked_into();
-                if !global.worker().is_undefined() {
-                    Some(Self::Worker(global.unchecked_into()))
-                } else {
-                    None
-                }
-            }
+        if let Some(window) = window() {
+            return Some(Self::Window(window));
         }
+
+        let global: Global = js_sys::global().unchecked_into();
+
+        if has(&global, "WorkerGlobalScope") {
+            return Some(Self::Worker(global.worker().unchecked_into()));
+        }
+
+        // check if the `setInterval` and `clearInterval` are available globally
+        if has(&global, "setInterval") && has(&global, "clearInterval") {
+            return Some(Self::Unknown(global));
+        }
+
+        None
+    }
+
+    pub(crate) fn is_websocket_supported() -> bool {
+        static SUPPORT: WebSocketSupport = WebSocketSupport::new();
+
+        SUPPORT.is_supported()
     }
 
     /// The `setInterval()` method.
-    pub(crate) fn set_interval_with_callback_and_timeout_and_arguments(
+    pub(crate) fn set_interval(
         &self,
-        handler: &::js_sys::Function,
+        handler: &js_sys::Function,
         timeout: i32,
-        arguments: &::js_sys::Array,
-    ) -> Result<i32, JsValue> {
+        arguments: &js_sys::Array,
+    ) -> Result<IntervalHandle, JsValue> {
         match self {
-            WebContext::Window(w) => {
-                w.set_interval_with_callback_and_timeout_and_arguments(handler, timeout, arguments)
-            }
-            WebContext::Worker(w) => {
-                w.set_interval_with_callback_and_timeout_and_arguments(handler, timeout, arguments)
-            }
+            Self::Window(window) => window
+                .set_interval_with_callback_and_timeout_and_arguments(handler, timeout, arguments)
+                .map(IntervalHandle::from_i32),
+            Self::Worker(worker) => worker
+                .set_interval_with_callback_and_timeout_and_arguments(handler, timeout, arguments)
+                .map(IntervalHandle::from_i32),
+            Self::Unknown(global) => global
+                .set_interval(handler, timeout, arguments)
+                .map(IntervalHandle::from_value),
         }
     }
 
     /// The `clearInterval()` method.
-    pub(crate) fn clear_interval_with_handle(&self, handle: i32) {
+    ///
+    /// # Panics
+    ///
+    /// This method panics if the handle given invalid.
+    /// This only happens whenever the handle was created by another
+    /// `WebContext` and that context has a different representation for the
+    /// handle.
+    ///
+    /// In particular this happens when one tries to call `clear_interval` on a
+    /// handle that was created in an unknown context in a known context.
+    pub(crate) fn clear_interval(&self, handle: IntervalHandle) {
         match self {
```

Comment:
I think this would be cleaner if you'd match on the tuple:

```rust
match (self, handle) {
	(Self::Window(window), IntervalHandle(HandleValue::Numeric(i))) => { ... },
	// ...
}
```

Reply:
not necessarily, we don't _know_ what the handle is and the `JsValue` returned might also be an `i32`, in that case it is totally valid to give it to the window or worker as well.

Reply:
Really? Isn't this error case somewhat impossible? The JS environment shouldn't suddenly change and return numbers instead of objects.

If we'd have dependent types then this would be type-safe, right?

Reply:
Yes via dependent types this would be type-safe, but just to be a bit more on the edge of caution. I could imagine an (unlikely) case where someone changes the globals, e.g., from the nodejs ones to the browser or vice versa; in that case, we wouldn't recover. While the possibility for this is not high, I thought if we can, we can build in some additional resiliency.

Reply:
Okay fair! I think it would still be cleaner to do something like the following then:

```rust
match (self, handle.into_i32()) {
	(Self::Window(window), Some(handle)) => {
		window.clear_interval_with_handle(handle);
	},
	// ...
}
```